Source/hydro/Castro_mol.cpp

```diff
@@ -6,10 +6,260 @@
+
+#ifdef GRAVITY
+  Real lconst_grav =  gravity->get_const_grav();
+#else
+  Real lconst_grav = 0.0_rt;
+#endif
+
+  AMREX_PARALLEL_FOR_4D(bx, NQ, i, j, k, n,
+  {
+
+   // piecewise linear slopes
+   uslope(bx, idir,
+          q_arr, n,
+          flatn_arr, dq);
+
+
+   // this is a loop over zones.  For each slope in the zone, fill the
+   // two adjacent edge states (e.g., the right state at i-1/2 and the
+   // left state at i+1/2
+
+   if (idir == 0) {
+
+     if (plm_well_balanced == 1 && n == QPRES && idir == AMREX_SPACEDIM-1) {
+
+       // left state at i+1/2 interface
+       qm(i+1,j,k,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n) +
+         0.5_rt*dx[0]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+       // right state at i-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n) -
+         0.5_rt*dx[0]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+     } else {
+       // left state at i+1/2 interface
+       qm(i+1,j,k,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n);
+
+       // right state at i-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n);
+
+     }
+
+#if AMREX_SPACEDIM >= 2
+   } else if (idir == 1) {
+
+     if (plm_well_balanced == 1 && n == QPRES && idir == AMREX_SPACEDIM-1) {
+
+       // left state at i+1/2 interface
+       qm(i,j+1,k,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n) +
+         0.5_rt*dx[1]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+       // right state at i-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n) -
+         0.5_rt*dx[1]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+     } else {
+
+       // left state at j+1/2 interface
+       qm(i,j+1,k,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n);
+
+       // right state at j-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n);
+
+     }
+#endif
+
+#if AMREX_SPACEDIM == 3
+   } else {
+
+     if (plm_well_balanced == 1 && n == QPRES && idir == AMREX_SPACEDIM-1) {
+
+       // left state at i+1/2 interface
+       qm(i,j,k+1,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n) +
+         0.5_rt*dx[2]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+       // right state at i-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n) -
+         0.5_rt*dx[2]*q_arr(i,j,k,QRHO)*lconst_grav;
+
+     } else {
+
+       // left state at k+1/2 interface
+       qm(i,j,k+1,n) = q_arr(i,j,k,n) + 0.5_rt*dq(i,j,k,n);
+
+       // right state at k-1/2 interface
+       qp(i,j,k,n) = q_arr(i,j,k,n) - 0.5_rt*dq(i,j,k,n);
+
+     }
+#endif
+   }
+
+  });
+
+
+  // special care for reflecting BCs
+  const int* lo_bc = phys_bc.lo();
+  const int* hi_bc = phys_bc.hi();
+
+  const auto domlo = geom.Domain().loVect3d();
+  const auto domhi = geom.Domain().hiVect3d();
+
+  bool lo_bc_test = lo_bc[idir] == Symmetry;
+  bool hi_bc_test = hi_bc[idir] == Symmetry;
+
+  // we have to do this after the loops above, since here we will
+  // consider interfaces, not zones
+
+  if (idir == 0) {
+    if (lo_bc_test) {
+
+      AMREX_PARALLEL_FOR_3D(bx, i, j, k,
```

Comment:
To match what we currently do, these loops have to be over a larger box that includes the ghost zones (grown by `dg[idir]` in each orthogonal direction).

Reply:
I think that was not necessary actually.  This box is already bigger than needed since it is the valid box + 1 zone in all directions, plus and minus.  I don't know why we need the extra buffer there for 2nd order.

Reply:
OK, I agree. And anyway the Fortran code was not right for CUDA, it looks like we missed that when we did the original port of that code.